### PR TITLE
RFC: 0.5 updates and redesign

### PR DESCRIPTION
**Pull request by tbreloff**
Ref: https://github.com/JuliaML/MLDataUtils.jl/issues/3#issuecomment-245670963

@Evizero I want you to be on board with this redesign.  I've already implemented random minibatches with a pattern:

``` julia
for batch in batches(inputs, targets, batch_size=100)
    ...
    for (input, target) in batch
        ...
    end
    ...
end
```

and I like the resulting usage.  I use it in StochasticOptimization.jl now: https://github.com/JuliaML/StochasticOptimization.jl/blob/master/test/runtests.jl#L37

What do you think?  Should I roll out similar updates for KFolds, etc?

### CODE REVIEW

**Comment by Evizero** [comment]
I am going to be straight with you. I am really unsure of this all. It starts with the completely different type tree and seeing KFolds as a DataIterator. I could get behind solving `RandomSamples` vs `LabeledRandomSamples` differently, or even maybe adding a third inner layer of having a "Data wrapper" in the inner loop. But at a first glance this seems like a completely different vision of how to perform data handling that I am not yet sure I share.

Let's maybe start with something that is easier communicated. When you look at code examples in the README at the moment. How would they look like with all your envisioned changes implemented?

File: src/dataiterators/dataiterator.jl [hunk 1]
```diff
@@ -1,30 +1,53 @@
...
+# apply a view to the last dimension
+@generated function LearnBase.getobs(A::AbstractArray, idx)
+    T, N = A.parameters
+    @assert N > 0
+    if N == 1 && idx <: Integer
+        :(A[idx])
+    else
+        :(view(A,  $(fill(:(:),N-1)...), idx))
+    end
+end
```
Comment (Evizero) [inline src/dataiterators/dataiterator.jl:37]: This does something different. I only used views for `Array` and `SubArray` and not for all `AbstractArray`. Is this change on purpose because of 0.5 improvements? What happens if it call it with a sparse array?

Comment (Evizero) [inline src/dataiterators/dataiterator.jl:37]: I do like the idea of using generated functions though

Comment (Evizero) [inline src/dataiterators/dataiterator.jl:37]: Also, I take it views don't require ranges anymore? What if I call this with `idx = [5, 3, 9]` for example; will `view` still work then?

File: src/dataiterators/dataiterator.jl [hunk 2]
```diff
@@ -1,30 +1,53 @@
...
+
+# add support for arbitrary tuples
+LearnBase.nobs{T<:Tuple}(tup::T) = nobs(tup[1])
+LearnBase.getobs{T<:Tuple}(tup::T, idx) = map(a -> getobs(a, idx), tup)
```
Comment (Evizero) [inline src/dataiterators/dataiterator.jl:41]: so map will return a `Tuple` here? I like it

File: src/dataiterators/minibatches.jl [hunk 3]
```diff
@@ -1,17 +1,16 @@
...
-function _compute_partitionsettings(features, size::Int, count::Int)
-    num_observations = nobs(features)::Int
+function _compute_partitionsettings(num_observations::Int, size::Int, count::Int)
     @assert num_observations > 0
     @assert size  <= num_observations
     @assert count <= num_observations
     if size < 0 && count < 0
         # no batch settings specified, use default size and as many batches as possible
-        size = default_partitionsize(features)::Int
+        size = default_partitionsize(num_observations)::Int
```
Comment (Evizero) [inline src/dataiterators/minibatches.jl:13]: Mhm, this was originally intended to give a user the possible entry point of controlling the default value for his/her custom datatype. Should we try to preserve that? My reason for that was that it is hard to know what kind of twisted way the data is loaded from; could be from a database, or network, or anything. So maybe one would want to define a sensible default partition for such a special case.

Comment (tbreloff) [inline src/dataiterators/minibatches.jl:13]: Thanks for all the comments. I might not have time to properly respond for
a couple days, but know that I'm reading them and appreciate it.

**Comment by tbreloff** [comment]
One point to make... there has been a lot of big improvements to the array view architecture, and I think it's a worthwhile goal to try to avoid copying and allocating new arrays whenever possible.  One can always throw in a `copy(x)` if they want a new array, but they can't go the other way.  Much of the change in logic is in using views instead of copying.

Aside from that, a core difference in the minibatches is that you have:

``` julia
for (batch_X, batch_y) in MiniBatches(train_X, train_y, size=10)
        # ... train supervised model here
    end
```

and I have:

``` julia
for batch in batches(x, y, batch_size = 10)
    ...
end
```

There are a couple of ways to resolve this... one is to get your logic with `get(batch)`:

``` julia
for batch in batches(x, y, batch_size = 10)
    batch_x, batch_y = get(batch)
    ...
end
```

or we could support both versions with different methods (the only difference is that one would call `get(batch)` inside `Base.next`.

Also to ease some of your general concerns about views, hopefully this helps show what you can do:

``` julia
julia> x = rand(10)
10-element Array{Float64,1}:
 0.666315
 0.416924
 0.38141
 0.561459
 0.213298
 0.665091
 0.428458
 0.368679
 0.132488
 0.591735

julia> x[[1,3,5]]
3-element Array{Float64,1}:
 0.666315
 0.38141
 0.213298

julia> view(x, [1,3,5])
3-element SubArray{Float64,1,Array{Float64,1},Tuple{Array{Int64,1}},false}:
 0.666315
 0.38141
 0.213298

julia> copy(ans)
3-element Array{Float64,1}:
 0.666315
 0.38141
 0.213298

julia> y = view(x, [1,3,5])
3-element SubArray{Float64,1,Array{Float64,1},Tuple{Array{Int64,1}},false}:
 0.666315
 0.38141
 0.213298

julia> z = view(y, 2:3)
2-element SubArray{Float64,1,Array{Float64,1},Tuple{Array{Int64,1}},false}:
 0.38141
 0.213298

julia> z[1] = 1000
1000

julia> x
10-element Array{Float64,1}:
    0.666315
    0.416924
 1000.0
    0.561459
    0.213298
    0.665091
    0.428458
    0.368679
    0.132488
    0.591735

julia> y
3-element SubArray{Float64,1,Array{Float64,1},Tuple{Array{Int64,1}},false}:
    0.666315
 1000.0
    0.213298

julia> z
2-element SubArray{Float64,1,Array{Float64,1},Tuple{Array{Int64,1}},false}:
 1000.0
    0.213298
```

tl;dr I think we can still support the examples from the readme, but under the hood we would replace "normal/labeled" versions with a single version that supports Tuples of data, and we would avoid copying and use views.

**Comment by tbreloff** [comment]
And a superficial point... I think the lowercase api looks nicer, and matches julia style more.  For example we do:

``` julia
for i in eachindex(x)
    ...
end
```

not

``` julia
for i in IndexIterator(x)
    ...
end
```

Again this is superficial and shouldn't distract from the core proposal which is to use more views and to simplify/generalize the code.

**Comment by Evizero** [comment]
Ok. Well, I agree on the view vs copy point.

You know, this example

``` julia
for (batch_X, batch_y) in MiniBatches(train_X, train_y, size=10)
        # ... train supervised model here
end
```

can also be written as

``` julia
for batch in MiniBatches(train_X, train_y, size=10)
        # ... train supervised model here
end
```

since `batch` is a tuple. I think this is a.) more general and b.) forces less things on the user as `batch_X` and `batch_y` are of the type the user expects instead of some custom data structure.

I do not see the benefit of making sure that `batch` is of the same type for supervised and unsupervised examples. For unsupervised it is the feature batch and for supervised it is a tuple of feature batch and target batch, as one would expect

**Comment by Evizero** [comment]
In other words, the main thing I see in your batches example is that it takes away the nice splicing option from the user.

**Comment by tbreloff** [comment]
> You know, this example ... can also be written as ... since batch is a tuple.

Yes, but then to be able to iterate through `batch` one observation at a time you'd need to re-wrap the tuple with an "observation iterator" of some sort.  You can't just do `for (x,y) in batch ...`like I wanted.

I think making a `Batch` type was a mistake... all it is is a `DataSubset`.  Likewise, test/train split and minibatches are just examples of "data partitions".  In my mind,
- `DataSubset` is a lightweight (non-copy) wrapper around one or more datasets
- `DataSubsets` is an iterator/view which returns some number of `DataSubset`
- test/train split and similar just return a `Tuple{DataSubset}`
- minibatches and similar are just special ways to construct a `DataSubsets`
- kfolds, LOOfolds, etc iterate by returning a `Tuple{DataSubset}` at each step (passing through params to test/train split)
- you could call `get` on a `DataSubset` to get the block views: `x, y = get(subset)` and there could be a way to choose how to iterate (do we call `get` automatically?)
- iterating over a `DataSubset` will give you a tuple of one observation for each underlying dataset

The net result is that the type tree is highly compressed and more general (allowing iteration over more than 2 datasets), and the functionality is just different ways to build these simple/common types.

If you're really uncomfortable with all this, I'll add it to StochasticOptimization instead (since no one else is using it yet) and leave MLDataUtils alone, and then we can compare side-by-side and have a discussion again in the future.

**Comment by tbreloff** [comment]
I'm adding a very simple version of this concept to StochasticOptimization so that you don't have to feel pressured into making drastic changes to MLDataUtils yet.  We can discuss again sometime in the future after I've experimented a little more.
